openstack_exporter/collectors/cinderbackend.py

```diff
@@ -43,40 +43,91 @@ def describe(self):
+        as reported by the backend.
+        total_reserved_available_capacity - The total capacity minus reserved
+        capacity
+
+        max_over_subscription_ratio - as reported by the backend
+        total_available_capacity - The total capacity available to cinder
+        calculated
+        thick: total_reserved_available_capacity
+        OR
+        thin: total_reserved_available_capacity and max_over_subscription_ratio
+
+        provisioned_capacity - as reported by backend or volume manager
+        (allocated_capacity_gb)
+
+        calculated_free_capacity - total_available_capacity - provisioned_capacity
+        virtual_free_capacity - The calculated free capacity available to cinder
+        to allocate new storage.
+        For thin: calculated_free_capacity
+        For thick: the reported free_capacity can be less than the calculated
+        Capacity, so we use free_capacity - reserved_capacity.
+
+        free_percent - the percentage of the total_available_capacity is left over
+        provisioned_ratio - The ratio of provisioned storage to
+        total_available_capacity
         """
 
         total = float(total_capacity)
         reserved = float(reserved_percentage) / 100
-        total_avail = total - math.floor(total * reserved)
+        reserved_capacity = math.floor(total * reserved)
+        total_avail = total - reserved_capacity
 
         if thin and thin_provisioning_support:
-            free = (total_avail * max_over_subscription_ratio - provisioned_capacity)
+            total_available_capacity = total_avail * max_over_subscription_ratio
+            calculated_free = total_available_capacity - provisioned_capacity
+            virtual_free = calculated_free
+            provisioned_type = 'thin'
         else:
             # Calculate how much free space is left after taking into
             # account the reserved space.
-            free = free_capacity - math.floor(total * reserved)
-        return free
+            total_available_capacity = total_avail
+            calculated_free = total_available_capacity - provisioned_capacity
+            virtual_free = calculated_free
+            if free_capacity < calculated_free:
+                virtual_free = free_capacity
+            provisioned_type = 'thick'
+
+        if total_available_capacity:
+            provisioned_ratio = provisioned_capacity / total_available_capacity
+            free_percent = (virtual_free / total_available_capacity) * 100
+        else:
+            provisioned_ratio = 0
+            free_percent = 0
+
+        return {
+            "total_capacity": total,
+            "free_capacity": free_capacity,
+            "reserved_capacity": reserved_capacity,
+            "total_reserved_available_capacity": int(total - reserved_capacity),
+            "max_over_subscription_ratio": (
+                max_over_subscription_ratio if provisioned_type == 'thin' else None
```

Comment:
Does it make sense to have `None` here or would `1` maybe serve a better purpose? That way,  one could maybe always use the `max_over_subscription_ratio` in computations, but it wouldn't change the outcome. Depends on how we use it though, so I leave the decision to you.

Reply:
I've gone back and forth on this too.   I had set it to 1 at one point, but if the provisioning type is thick max_over_subscription_ratio should never be used.   So, that's why I settled on None.